File: reference/mic-cam-switcher-node/src/audio/atemFairlightSource.js
const { EventEmitter } = require('events');

/**
 * EXPERIMENTAL. Blackmagic's ATEM protocol is reverse-engineered, and how much
 * Fairlight audio meter data it exposes over the network varies by model/firmware.
 * atem-connection surfaces Fairlight mixer *state* (gain, EQ, dynamics) reliably,
 * but live peak/RMS meter ticks are not consistently available across all ATEM
 * hardware. This polls whatever level data atem-connection's state tree exposes
 * for each Fairlight input source and converts it to an approximate dB value.
 *
 * If your ATEM doesn't report live levels this way, the practical workaround is
 * to also tap the same mic feeds into the Sound Devices interface (a split or a
 * second output) and use SystemAudioSource instead - same UI, same switch engine,
 * just a different level source per mic.
 */
class AtemFairlightSource extends EventEmitter {
  constructor(atemController, { pollMs = 50 } = {}) {
    super();
    this.atemController = atemController;
    this.pollMs = pollMs;
    this.timer = null;
  }

  start() {
    this.timer = setInterval(() => this._poll(), this.pollMs);
  }

  stop() {
    if (this.timer) clearInterval(this.timer);
    this.timer = null;
  }

  _poll() {
    const atem = this.atemController.atem;
    if (!atem || !this.atemController.connected) return;

    const fairlight = atem.state && atem.state.fairlight;
    if (!fairlight || !fairlight.inputs) {
      this.emit('unavailable');
      return;
    }

    // Map of atemInputId -> approximate dB level, keyed by Fairlight input index.
    const levels = {};
    for (const [inputId, input] of Object.entries(fairlight.inputs)) {
      const sources = input.sources || {};
      const firstSource = Object.values(sources)[0];
      // faderGain is a mix-level setting, not a live meter - this is a placeholder
      // until confirmed against real hardware. Treat AtemFairlightSource as a
      // starting point to wire up against your actual switcher, not a finished part.
      const gain = firstSource && typeof firstSource.faderGain === 'number' ? firstSource.faderGain : -100;
      levels[inputId] = gain;
    }

    this.emit('levels', levels);
  }
}

module.exports = AtemFairlightSource;

File: app/audio/atem_fairlight_source.py
# app/audio/atem_fairlight_source.py
import asyncio

"""
EXPERIMENTAL. See docs/superpowers/specs/2026-06-30-mic-cam-atem-switcher-design.md,
"Deferred / experimental" section. PyATEMMax exposes Fairlight mixer *state* (gain,
EQ, dynamics) but not consistent live meter ticks across ATEM firmware/models. Kept
as a documented starting point -- confirm the state-tree attribute names below
against your actual switcher's PyATEMMax object before relying on this; the reliable
path is tapping mics into the audio interface and using system_audio_source.py.
"""


class AtemFairlightSource:
    def __init__(self, atem_controller, poll_ms=50):
        self.atem_controller = atem_controller
        self.poll_ms = poll_ms
        self._on_levels_cb = None
        self._on_unavailable_cb = None

    def on_levels(self, callback):
        self._on_levels_cb = callback

    def on_unavailable(self, callback):
        self._on_unavailable_cb = callback

    async def start(self):
        while True:
            self._poll()
            await asyncio.sleep(self.poll_ms / 1000)

    def _poll(self):
        if not self.atem_controller.connected:
            return
        fairlight_state = getattr(self.atem_controller.switcher, "fairlightAudioMixerInput", None)
        if not fairlight_state:
            if self._on_unavailable_cb:
                self._on_unavailable_cb()
            return
        # Unconfirmed against real hardware: PyATEMMax's Fairlight state-tree layout
        # and attribute names need verification (see module docstring above).
        levels = {}
        if self._on_levels_cb:
            self._on_levels_cb(levels)
